README.md
# ahnr <img src="man/figures/AON_molecule.png" align="right" width="25%"/>
[![Travis-CI Build Status](https://travis-ci.org/jroberayalas/ahnr.svg?branch=master)](https://travis-ci.org/jroberayalas/ahnr)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/ahnr)](https://CRAN.R-project.org/package=ahnr)

`ahnr` is a package that implements the artificial hydrocarbon networks developed by [Hiram Ponce](http://www.solariumlabs.com/hiram_ponce.html). Most of the work is based on the book [Artificial Organic Networks](http://www.springer.com/gb/book/9783319024714). You can see the package in action at https://jroberayalas.github.io/ahnr/

## Installation

You can install the `ahnr` package from CRAN:

```R
install.packages("ahnr")
```

The development version can be installed from GitHub using the `devtools` package:

```R
install.packages("devtools")
devtools::install_github('jroberayalas/ahnr')
```
## Quick Start

For a quick start with the package, check the [vignette](https://jroberayalas.github.io/ahnr/articles/ahn.html).


NEWS.md
# ahnr 0.3.2

* [Issue 2](https://github.com/jroberayalas/ahnr/issues/2) (fit function) solved.
* [Issue 3](https://github.com/jroberayalas/ahnr/issues/3) (relocate molecules) solved.

# ahnr 0.3.1

* Removing generics.
* plot function name changed to visualize.

# ahnr 0.3.0

* Generic functions updated: summary, predict, plot.
* AHNnD function name changed to fit.
* SimAHNnD function name changed to predict.
* Fix in `predict` function: ncol(H[[1]]) --> max(unlist(sapply(H, ncol)))
* Fix in `predict` function to handle empty molecules: if (is.null(H[[i]])) {next}

# ahnr 0.2.1

* Correction to the labels of the second plot in the vignette.

# ahnr 0.2.0

* Added legends in the plots of the vignette file.
* Correction due to the changes to the R development sources: "arithmetic with zero-column data.frames".

# ahnr 0.1.0

* Added a `NEWS.md` file to track changes to the package.



R/generics.R
#' Summary Artificial Hydrocarbon Network
#'
#' @description Summary method for objects of class \code{ahn}.
#'
#' @param x an object of class "\code{ahn}" produced from the \link{fit} function.
#' @param ... further arguments passed to or from other methods.
#'
#' @return summary description of the AHN.
#' @export
#'
#' @examples
#' \dontrun{
#' # Create data
#' x <- 2 * runif(1000) - 1;
#' x <- sort(x)
#'
#' y <- (x < 0.1) * (0.05 * runif(100) + atan(pi*x)) +
#'     (x >= 0.1 & x < 0.6) * (0.05 * runif(1000) + sin(pi*x)) +
#'     (x >= 0.6) * (0.05 * runif(1000) + cos(pi*x))
#'
#' # Create Sigma list
#' Sigma <- list(X = data.frame(x = x), Y = data.frame(y = y))
#'
#' # Train AHN
#' ahn <- fit(Sigma, 5, 0.01, 500)
#'
#' # Summary AHN
#' summary(ahn)
#' }
#'
#summary <- function(x, ...) {
#    UseMethod("summary", x)
#}


#' predict
#'
#' @description Function to simulate a trained Artificial Hydrocarbon Network.
#'
#' @param x an object of class "\code{ahn}" produced from the \link{fit} function.
#' @param ... further arguments passed to or from other methods.
#'
#' @return predicted output values for inputs in \code{new_data}.
#' @export
#'
#' @examples
#' \dontrun{
#' # Create data
#' x <- 2 * runif(1000) - 1;
#' x <- sort(x)
#'
#' y <- (x < 0.1) * (0.05 * runif(100) + atan(pi*x)) +
#'     (x >= 0.1 & x < 0.6) * (0.05 * runif(1000) + sin(pi*x)) +
#'     (x >= 0.6) * (0.05 * runif(1000) + cos(pi*x))
#'
#' # Create Sigma list
#' Sigma <- list(X = data.frame(x = x), Y = data.frame(y = y))
#'
#' # Train AHN
#' ahn <- fit(Sigma, 5, 0.01, 500)
#'
#' # Test AHN
#' X <- data.frame(x = x)
#' ysim <- predict(ahn, X)
#' }
#'
#predict <- function(x, ...) {
#    UseMethod("predict", x)
#}


#' Plot Artificial Hydrocarbon Network
#'
#' @description Plot method for objects of class \code{ahn}.
#'
#' @param x an object of class "\code{ahn}" produced from the \link{fit} function.
#' @param ... further arguments passed to visNetwork functions.
#'
#' @return dynamic visualization of the AHN.
#' @export
#'
#' @examples
#' \dontrun{
#' # Create data
#' x <- 2 * runif(1000) - 1;
#' x <- sort(x)
#'
#' y <- (x < 0.1) * (0.05 * runif(100) + atan(pi*x)) +
#'     (x >= 0.1 & x < 0.6) * (0.05 * runif(1000) + sin(pi*x)) +
#'     (x >= 0.6) * (0.05 * runif(1000) + cos(pi*x))
#'
#' # Create Sigma list
#' Sigma <- list(X = data.frame(x = x), Y = data.frame(y = y))
#'
#' # Train AHN
#' ahn <- fit(Sigma, 5, 0.01, 500)
#'
#' # Plot AHN
#' plot(ahn)
#' }
#'
#plot <- function(x, ...) {
#    UseMethod("plot", x)
#}


R/auxiliary.R
# CH_X creates the Vandermonde matrix
CH_X <- function(X, h) {
    N <- ncol(X)
    q <- nrow(X)

    Phi <- matrix(0, nrow = q, ncol = N * (h+1))

    for (i in seq_len(N)) {
        #Phi <- cbind(Phi, matrixcalc::vandermonde.matrix(X[ , i], h+1))
        Phi[ , (1:(h+1)) + (i-1)*(h+1)] <- matrixcalc::vandermonde.matrix(X[ , i], h+1)
    }
    #Phi <- Phi[ , -1]

    if (is.vector(Phi)) {
        Phi <- matrix(Phi, nrow = 1)
    }

    Phi
}

# ComputeMoleculeParameters solves the system of equations with Least Squares
ComputeMoleculeParameters <- function(Xi, Yi, Omegai) {
    Phi <- CH_X(Xi, Omegai)

    H <- pracma::mldivide(Phi, as.matrix(Yi))
    Yapprox <- Phi %*% H

    errorMolecule <- sqrt(mean(abs((Yapprox - as.matrix(Yi))^2)))

    list(H = H, Yapprox = Yapprox, errorMolecule = errorMolecule)
}

# CreateLinearCompound creates a network with n molecules
CreateLinearCompound <- function(n) {
    Omega <- 2 * rep(1, n)
    Omega[1] <- 3
    Omega[length(Omega)] <- 3

    B <- rep(1, n-1)

    list(Omega = Omega, B = B)
}

# InitialPosition defines the initial positions of the n molecules
InitialPosition <- function(X, n) {
    minValues <- purrr::map_dbl(X, min)
    maxValues <- purrr::map_dbl(X, max)

    ranges <- maxValues - minValues

    rangesMatrix <- matrix(ranges, nrow = n, ncol = length(ranges), byrow = TRUE)
    rUnifMatrix <- matrix(runif(n*length(ranges)), nrow = n, ncol = length(ranges), byrow = TRUE)
    minMatrix <- matrix(minValues, nrow = n, ncol = length(ranges), byrow = TRUE)
    posMolecules <- rangesMatrix * rUnifMatrix + minMatrix

    ###
    # rangesMatrix[1, ] <- minValues + (ranges / (n * 2))
    # for (k in seq_len(n-1)) {
    #     rangesMatrix[(k+1), ] <- minValues + (ranges * (2 * k + 1)/ (n * 2))
    # }
    # posMolecules <- rangesMatrix
    ###

    posMolecules
}

# DataInMolecules groups the data to the closer molecule
DataInMolecules <- function(Sigma, posMolecules) {
    X <- Sigma$X
    Y <- Sigma$Y

    numSamples <- nrow(X)
    n <- nrow(posMolecules)

    # dist <- matrix(0, nrow = numSamples, ncol = n)
    #
    # for (i in seq_len(n)) {
    #     for (j in seq_len(numSamples)) {
    #         dist[j, i] <- Distance(X[j, ], posMolecules[i, ])
    #     }
    # }

    dist <- as.matrix(pdist::pdist(as.matrix(X), posMolecules))

    moleculeNumber <- factor(paste("molecule", apply(dist, 1, which.min), sep = ""))

    SigmaSplitX <- split(X, moleculeNumber)
    SigmaSplitY <- split(Y, moleculeNumber)
    SigmaSplit <- list(X = SigmaSplitX, Y = SigmaSplitY)
    SigmaSplit

    list(SigmaSplit = SigmaSplit, moleculesUsed = sort(unique(apply(dist, 1, which.min))), moleculeNumber = moleculeNumber)
}

# RelocateMolecules relocates molecules with no data
RelocateMolecules <- function(posMolecules, moleculesUsed, errorMolecule) {
    n <- nrow(posMolecules)

    errorMolecules <- errorMolecule[moleculesUsed, , drop = FALSE]
    moleculeNumber <- sort(apply(errorMolecules, 1, sum), decreasing = TRUE, index.return = TRUE)$ix

    seekMolecule <- 1

    for (i in seq_len(n)) {
        if (!(i %in% moleculesUsed)) {
            radius <- 1
            posMolecules[i, ] <- rep(radius, ncol(posMolecules)) * runif(ncol(posMolecules)) + posMolecules[moleculeNumber[seekMolecule], ]

            seekMolecule <- seekMolecule + 1

            if (seekMolecule > length(moleculeNumber)) {
                seekMolecule <- 1
            }
        }
    }

    requireRelocation <- !(nrow(errorMolecules) == n)

    list(requireRelocation = requireRelocation, posMolecules = posMolecules)
}

# SimDataInMolecules groups the data to the closer molecule based on the trained network
SimDataInMolecules <- function(X, posMolecules) {
    numSamples <- nrow(X)
    n <- nrow(posMolecules)

    index <- 1:numSamples

    # dist <- matrix(0, nrow = numSamples, ncol = n)
    #
    # for (i in seq_len(n)) {
    #     for (j in seq_len(numSamples)) {
    #         dist[j, i] <- Distance(X[j, ], posMolecules[i, ])
    #     }
    # }

    dist <- as.matrix(pdist::pdist(as.matrix(X), posMolecules))

    moleculeNumber <- factor(paste("molecule", apply(dist, 1, which.min), sep = ""))

    SigmaSplitX <- split(X, moleculeNumber)
    SigmaSplitIndex <- split(index, moleculeNumber)
    SigmaSplit <- list(X = SigmaSplitX, Index = SigmaSplitIndex)
    SigmaSplit
}


# CreateLabels creates the labels for the network visualization
CreateLabels <- function(molecule, dimensions) {
    molecule_split <- split(molecule, ceiling(seq_len(length(molecule)) / (length(molecule) / dimensions)))
    do.call(paste, c(molecule_split, sep = "\n"))
}

# CreateNodes creates the data frame for the nodes of the network
CreateNodesEdges <- function(ahn) {
    label <- vector('numeric')
    id <- vector('character')
    group <- vector('character')
    from <- paste('C', seq_len(ahn$network$n - 1), sep = '')
    to <- paste('C', seq_len(ahn$network$n - 1) + 1, sep = '')
    for (i in seq_len(ahn$network$n)) {
        label <- c(label, CreateLabels(round(ahn$network$H[[paste('molecule', i, sep = "")]][ , 1], 3), ncol(ahn$network$Pi)))

        carbon <- paste('C', i, sep = '')
        hydrogens <- paste('H', i, seq_len(ahn$network$C$Omega[i]), sep = '')

        id <- c(id, carbon, hydrogens)

        group <- c(group, "C", paste('H', seq_len(ahn$network$C$Omega[i]), sep = ''))

        from <- c(from, rep(carbon, ahn$network$C$Omega[i]))
        to <- c(to, hydrogens)
    }

    nodes <- data.frame(id = id,
                        shape = 'circle',
                        shadow = FALSE,
                        label = label,
                        group = group,
                        font.color =c ("red", rep("black", length(id)-1)))

    edges <- data.frame(from = from,
                        to = to)

    list(nodes = nodes, edges = edges)
}

CreateTable <- function(ahn) {
    for (i in seq_len(ahn$network$n)) {
        cat("Molecule ", i,":\n", sep = "")

        molecule <- round(ahn$network$H[[paste('molecule', i, sep = "")]][ , 1], 3)
        dimensions <- ncol(ahn$network$Pi)
        molecule_split <- split(molecule, ceiling(seq_len(length(molecule)) / (length(molecule) / dimensions)))
        names(molecule_split) <- ahn$variableNames
        molecule_table <- as.data.frame(molecule_split)

        carbon <- paste('C', i, sep = '')
        hydrogens <- paste('H', i, seq_len(ahn$network$C$Omega[i]), sep = '')
        rownames(molecule_table) <- c(carbon, hydrogens)

        print(molecule_table)
        cat("\n")
    }
}


R/ahn_main.R
#' fit
#'
#' @description Function to train an Artificial Hydrocarbon Network (AHN).
#'
#' @param Sigma a list with two data frames. One for the inputs X, and one for the outputs Y.
#' @param n number of particles to use.
#' @param eta learning rate of the algorithm. Default is \code{0.01}.
#' @param maxIter maximum number of iterations.
#'
#' @return an object of class "\code{ahn}" with the following components:
#' \itemize{
#'         \item network: structure of the AHN trained. It contains the hydrogen coefficient H,
#'         position of the molecules Pi, data cluster ID per molecule moleculeNumber, number of molecules n,
#'         and carbon structure C
#'         \item Yo: original output variable.
#'         \item Ym: predicted output variable.
#'         \item eta: learning rate.
#'         \item minOverallError: minimum error achieved.
#'         \item variableNames: names of the input variables.
#' }
#' @export
#'
#' @examples
#' # Create data
#' x <- 2 * runif(1000) - 1;
#' x <- sort(x)
#'
#' y <- (x < 0.1) * (0.05 * runif(100) + atan(pi*x)) +
#'     (x >= 0.1 & x < 0.6) * (0.05 * runif(1000) + sin(pi*x)) +
#'     (x >= 0.6) * (0.05 * runif(1000) + cos(pi*x))
#'
#' # Create Sigma list
#' Sigma <- list(X = data.frame(x = x), Y = data.frame(y = y))
#'
#' # Train AHN
#' ahn <- fit(Sigma, 5, 0.01, 500)
#'
fit <- function(Sigma, n, eta, maxIter = 2000) {
    # Security Checking
    if (length(Sigma) != 2) {
        stop("Sigma must be a list with 2 data frames: one for the predictors and one for the outcome variables. ", call. = FALSE)
    }

    if (!is.data.frame(Sigma$X)) {
        stop("The first component of Sigma must be a data frame with the predictor variables. ", call. = FALSE)
    }

    if (!is.data.frame(Sigma$Y)) {
        stop("The second component of Sigma must be a data frame with the outcome variables. ", call. = FALSE)
    }

    if (n < 1) {
        stop("At least two particles are required.", call. = FALSE)
    }

    if ((eta <= 0) | (eta >= 1)) {
        stop("The learning rate eta must be between 0 and 1 (exclusive).", call. = FALSE)
    }

    if (maxIter < 1) {
        stop("Maximum number of iterations must be a positive integer.", call. = FALSE)
    }

    maxIter <- maxIter
    iter <- 1

    X <- Sigma$X
    Y <- Sigma$Y

    newIdx <- sort(Y[ , 1], index.return = TRUE)$ix
    X <- X[newIdx, , drop = FALSE]
    Y <- Y[newIdx, , drop = FALSE]

    numOutputs <- ncol(Y)

    overallError <- rep(Inf, maxIter)
    minOverallError <- Inf

    # Create a saturated and linear compound
    C <- CreateLinearCompound(n)

    # Initialize randomly the position of molecules
    posMolecules <- ahn_Pi <- InitialPosition(X, n)

    # Initialize AHN-structure
    H.names <- paste("molecule", 1:n, sep = "")
    H <- vector("list", length(H.names))
    names(H) <- H.names
    ahn_H <- H

    # Train compound
    while (iter < maxIter) {
        requireRelocation <- TRUE
        relocationIter <- 1
        maxRI <- 10

        while (requireRelocation & (relocationIter <= maxRI)) {
            molecules <- DataInMolecules(Sigma, posMolecules)

            Yo <- Ym <- matrix(0, nrow = nrow(Y), ncol = ncol(Y))
            pointer <- 1

            errorMolecule <- matrix(0, nrow = n+1, ncol = numOutputs)

            for (i in seq_len(length(molecules$SigmaSplit$X))) {
                moleculeUsed <- paste('molecule', molecules$moleculesUsed[i], sep = '')
                Xi <- molecules$SigmaSplit$X[[moleculeUsed]]
                Yi <- molecules$SigmaSplit$Y[[moleculeUsed]]

                parameters <- ComputeMoleculeParameters(Xi, Yi, C$Omega[molecules$moleculesUsed[i]])

                H[[moleculeUsed]] <- parameters$H

                errorMolecule[molecules$moleculesUsed[i], ] <- parameters$errorMolecule

                pointerNew <- pointer + nrow(parameters$Yapprox)
                Yo[pointer:(pointerNew-1), ] <- Yi[seq_len(nrow(Yi)), ]
                Ym[pointer:(pointerNew-1), ] <- parameters$Yapprox
                pointer <- pointerNew
            }

            relocation <- RelocateMolecules(posMolecules, molecules$moleculesUsed, errorMolecule)
            requireRelocation <- relocation$requireRelocation
            posMolecules <- relocation$posMolecules
            relocationIter <- relocationIter + 1
        }

        if (relocationIter >= maxIter) {
            H <- ahn_H
            posMolecules <- ahn_Pi
            moleculeNumber <- molecules$moleculeNumber
        }

        # Update AHN-structure
        overallError[iter] <- sum(errorMolecule)

        if (overallError[iter] < minOverallError) {
            ahn_H <- H
            ahn_Pi <- posMolecules
            moleculeNumber <- molecules$moleculeNumber
            minOverallError = overallError[iter]
        }

        # Update intermolecular distances
        for (i in seq_len(n)) {
            deltaPosMolecule <- -eta * (sum(errorMolecule[i, ]) - sum(errorMolecule[i+1, ])) * rep(1, ncol(posMolecules))
            posMolecules[i, ] <- posMolecules[i, ] + deltaPosMolecule
        }

        iter <- iter + 1
    }
    network = list(H = ahn_H, Pi = ahn_Pi, moleculeNumber = moleculeNumber, n = n, C = C)
    ahn <- list(network = network,
                Yo = Yo,
                Ym = Ym,
                eta = eta,
                minOverallError = minOverallError,
                variableNames = names(Sigma$X))
    class(ahn) <- "ahn"
    ahn
}


#' Checks if argument is a \code{ahn} object
#'
#' @param x An \R object
#'
#' @export
#'
is.ahn <- function(x) inherits(x, "ahn")


# #' plotAHN
# #'
# #' @param ahn a list produced from the \link{fit} function.
# #'
# #' @return visualization of the AHN.
# #' @export
# #'
# #' @examples
# #' # Create data
# #' x <- 2 * runif(1000) - 1;
# #' x <- sort(x)
# #'
# #' y <- (x < 0.1) * (0.05 * runif(100) + atan(pi*x)) +
# #'     (x >= 0.1 & x < 0.6) * (0.05 * runif(1000) + sin(pi*x)) +
# #'     (x >= 0.6) * (0.05 * runif(1000) + cos(pi*x))
# #'
# #' # Create Sigma list
# #' Sigma <- list(X = data.frame(x = x), Y = data.frame(y = y))
# #'
# #' # Train AHN
# #' ahn <- fit(Sigma, 5, 0.01, 500)
# #'
# #' # Plot AHN
# #' plotAHN(ahn)
# #'
# plotAHN <- function(ahn) {
#     vis <- CreateNodesEdges(ahn)
#     graph <- igraph::graph_from_data_frame(vis$edges, directed = FALSE, vertices = vis$nodes)
#     ggraph(graph, layout = 'graphopt') +
#         geom_edge_link(aes(start_cap = label_rect(node1.name),
#                            end_cap = label_rect(node2.name))) +
#         geom_node_text(label = vis$nodes$label) +
#         theme_void()
# }


#' Summary Artificial Hydrocarbon Network
#'
#' @description Summary method for objects of class \code{ahn}.
#'
#' @param object an object of class "\code{ahn}" produced from the \link{fit} function.
#' @param ... further arguments passed to or from other methods.
#'
#' @return summary description of the AHN.
#' @export
#'
#' @examples
#' \dontrun{
#' # Create data
#' x <- 2 * runif(1000) - 1;
#' x <- sort(x)
#'
#' y <- (x < 0.1) * (0.05 * runif(100) + atan(pi*x)) +
#'     (x >= 0.1 & x < 0.6) * (0.05 * runif(1000) + sin(pi*x)) +
#'     (x >= 0.6) * (0.05 * runif(1000) + cos(pi*x))
#'
#' # Create Sigma list
#' Sigma <- list(X = data.frame(x = x), Y = data.frame(y = y))
#'
#' # Train AHN
#' ahn <- fit(Sigma, 5, 0.01, 500)
#'
#' # Summary AHN
#' summary(ahn)
#' }
#'
summary.ahn <- function(object, ...) {
    stopifnot(is.ahn(object))
    ahn <- object

    cat("\nArtificial Hydrocarbon Network trained:\n\n")
    cat("Number of molecules:\n", ahn$network$n, "\n\n")
    cat("Learning factor:\n", ahn$eta, "\n\n")
    cat("Overall error:\n", round(ahn$minOverallError, 4), "\n\n")

    centers <- ahn$network$Pi
    rownames(centers) <- paste('molecule', seq_len(nrow(centers)), sep = "")
    colnames(centers) <- ahn$variableNames
    cat("Centers of the molecules:\n")
    print(as.table(centers))

    cat("\nMolecules:\n")
    CreateTable(ahn)
}


#' predict
#'
#' @description Function to simulate a trained Artificial Hydrocarbon Network.
#'
#' @param object an object of class "\code{ahn}" produced from the \link{fit} function.
#' @param ... further arguments passed to or from other methods.
#'
#' @return predicted output values for inputs in \code{newdata}.
#' @export
#'
#' @examples
#' \dontrun{
#' # Create data
#' x <- 2 * runif(1000) - 1;
#' x <- sort(x)
#'
#' y <- (x < 0.1) * (0.05 * runif(100) + atan(pi*x)) +
#'     (x >= 0.1 & x < 0.6) * (0.05 * runif(1000) + sin(pi*x)) +
#'     (x >= 0.6) * (0.05 * runif(1000) + cos(pi*x))
#'
#' # Create Sigma list
#' Sigma <- list(X = data.frame(x = x), Y = data.frame(y = y))
#'
#' # Train AHN
#' ahn <- fit(Sigma, 5, 0.01, 500)
#'
#' # Test AHN
#' X <- data.frame(x = x)
#' ysim <- predict(ahn, X)
#' }
#'
predict.ahn <- function(object, ...) {
    # Security Checking
    stopifnot(is.ahn(object))
    ahn <- object

    dots <- list(...)
    newdata <- dots[[1]]

    if (!is.data.frame(newdata)) {
        stop("newdata must be a data frame with the predictor variables. ", call. = FALSE)
    }

    # Extract network components
    H <- ahn$network$H
    posMolecules <- ahn$network$Pi
    C <- ahn$network$C

    # Initial statemets
    Yapprox <- matrix(0, nrow = nrow(newdata), ncol = max(unlist(sapply(H, ncol))))
    indexes <- rep(0, nrow(newdata))

    # Distribute data over molecules
    molecules <- SimDataInMolecules(newdata, posMolecules)

    # Evaluate AHN-model
    pointer <- 1

    for (i in seq_len(length(molecules$X))) {
        Xi <- molecules$X[[i]]
        indexesi <- molecules$Index[[i]]

        ki <- C$Omega[i]
        Phi <- CH_X(Xi, ki)

        if (is.null(H[[i]])) {next}

        Ym <- Phi %*% H[[i]]

        pointerNew <- pointer + nrow(Ym)
        Yapprox[pointer:(pointerNew-1), ] <- Ym#[seq_len(nrow(Ym)), ]
        indexes[pointer:(pointerNew-1)] <- indexesi
        pointer <- pointerNew
    }

    # Sort data
    ix <- sort(indexes, index.return = TRUE)$ix;
    Y <- Yapprox[ix, ]
}


#' Visualize Artificial Hydrocarbon Network
#'
#' @description Visualize method for objects of class \code{ahn}.
#'
#' @param x an object of class "\code{ahn}" produced from the \link{fit} function.
#' @param ... further arguments passed to visNetwork functions.
#'
#' @return dynamic visualization of the AHN.
#' @export
#'
#' @examples
#' \dontrun{
#' # Create data
#' x <- 2 * runif(1000) - 1;
#' x <- sort(x)
#'
#' y <- (x < 0.1) * (0.05 * runif(100) + atan(pi*x)) +
#'     (x >= 0.1 & x < 0.6) * (0.05 * runif(1000) + sin(pi*x)) +
#'     (x >= 0.6) * (0.05 * runif(1000) + cos(pi*x))
#'
#' # Create Sigma list
#' Sigma <- list(X = data.frame(x = x), Y = data.frame(y = y))
#'
#' # Train AHN
#' ahn <- fit(Sigma, 5, 0.01, 500)
#'
#' # Visualize AHN
#' visualize(ahn)
#' }
#'
visualize <- function(x, ...) {
    stopifnot(is.ahn(x))
    vis <- CreateNodesEdges(x)
    visNetwork(vis$nodes, vis$edges, width = "100%", ...) %>%
        visGroups(groupname = "C", color = "#fbb4ae", ...) %>%
        visGroups(groupname = "H1", color = "#b3cde3", ...) %>%
        visGroups(groupname = "H2", color = "#ccebc5", ...) %>%
        visGroups(groupname = "H3", color = "#decbe4", ...) %>%
        visLegend(position = "right", main = "Legend", ...)
}
